---
title: "Analysis of prostate cancer single-cell data"
subtitle: "Prostate cancer single-cell analysis"
author: "Bionformatics Service - Genomics Unit @ GENyO Research Center (Granada, Spain) bioinfo@genyo.es"
date: "`r Sys.Date()`"
output: 
 # html_notebook:
    toc: yes
---

# ABOUT.

The analysis in this document have been performed by the Bioinformatics Service of the Genomics Unit at GENyO Research Center (Granada, Spain).

For contact, please check the email below.


--------------------------------------------------------------------------------
Bioinformatics Service - Genomics Unit.

GENYO. Pfizer-University of Granada-Junta de Andalucia Center for Genomics and Oncological Research.

Parque Tecnológico de Ciencias de la Salud Avenida de la Ilustración 114 (18016).

Granada, Spain.

Phone: 958 715 500

Fax: +34 958 637 071

Email: [bioinfo\@genyo.es](mailto:bioinfo@genyo.es){.email}

Web: www.genyo.es

# LICENSE.

This markdown and all its content is published and distributed under a GPLv3 license: <https://www.gnu.org/licenses/gpl-3.0.html>.

```{r include = FALSE}

all_times <- list() # store the time for each chunk
knitr::knit_hooks$set(time_it = local({
  now <- NULL
  function(before, options) {
    if (before) {
      now <<- Sys.time()
    } else {
      res <- difftime(Sys.time(), now, units = "secs")
      all_times[[options$label]] <<- res
    }
  }
}))

knitr::opts_chunk$set(
  tidy = TRUE,
  tidy.opts = list(width.cutoff = 95),
  message = FALSE,
  warning = FALSE,
  time_it = TRUE,
  error = TRUE,
  echo = TRUE,
  engine.opts = list(bash = "-l")
)
```

```{r libraries, warning = FALSE, message = FALSE}

suppressMessages(library(Seurat))
suppressMessages(library(dplyr))
suppressMessages(library(ggplot2))
suppressMessages(library(ggpubr))
suppressMessages(library(celldex)) # Cell annotation.
suppressMessages(library(SingleR)) # Cell annotation.
suppressMessages(library(future)) # Allows parallelization in Seurat.
# Set up Seurat pararell computing.
options(parallelly.fork.enable = TRUE)
plan("multicore", workers = parallel::detectCores())
options(future.globals.maxSize = 8000 * 1024^2)

# Import user's libraries.
source(file = "~/Documents/SC_Prostate/Scripts/SC_utils/FeatureScatterGradient.R")
source(file = "~/Documents/SC_Prostate/Scripts/SC_utils/CalculateCellularDetectionRate.R")
source(file = "~/Documents/SC_Prostate/Scripts/SC_utils/scGSEAmarkers.R")

set.seed(0)
```

# Introduction.

Single-cell data was generated and pre-analysed sing BD Rhapsody systems and Illumina sequencers. BD includes pipelines for read mapping and putative cell filtering. The QC and filtering steps for each of the 4 sequencing batches are in a separate markdown. This document presents the analysis of the combined dataset.

# 1. Load filtered run data.

First, we search trough the project directory for each filtered raw data and load each SeuratObject.

```{r get_runs}

# List the data directory.
run_dirs <- list.dirs("../Data",recursive = FALSE)

# Filter only the run folders.
# Grepl stands for "logical grep" so it only outputs TRUE or FALSE. It accepts an array of strings as an input, search for teh given pattern and returns another array with TRUE or FALSE if the pattern matches with the string in that position of the array.
run_dirs <- run_dirs[grepl("run",run_dirs)]

# lapply stands for "list apply", and allows to apply a certain function over all elements of an array: lapply(X, FUN). You can apply an R-defined function or create you own as shown below.
rds_objects <- lapply(X = run_dirs, FUN = function(x) {

  rds_file <- list.files(path = x, pattern = "*_filtered.rds",full.names = T) # For each run directory list the filtered RDS object.

  # If there is something in the directory list...
    if (length(rds_file) > 0){
      rds_object <- readRDS(rds_file) # Load the RDS object in to a list.
      rds_object@meta.data$orig.ident <- basename(x) # Add the original path in the metadata to keep track of data.
      return(rds_object) # Return the list of RDS object and...

    } else {
      return(NULL) # If the directory list is empty return null.
    }
  }
)

# sapply is similar to lapply but returns a vector of the same lenght instead of the same element as the input (i.e. lapply returns a list is input is a list, but sapply always returns a vector).
rds_objects[sapply(rds_objects, is.null)] <- NULL # Filter emtpy objects.
```

# 2. Merge Seurat Objects.

We merge the 4 datasets into a unique Seurat Object. We confirm the merge by checking the total number off cells.

```{r merge_runs, fig.height=6, fig.width=10}

# Reduce() allows to reduce a list of values into a single value by applying a binary operation over then iteratively. That is, takes the first two elements of a list, x and y, apply a the function (fun(x,y)) creating the reduced value a, then takes the third element of the list, z, and apply the function to c and z: fun(c, z) and so on till the whole list is reduce to one single value. In this case, we apply Seruat's merge to merge all objects sequentially into one single SeuratObject for the whole data set.
# add.cell.ids = add the project name to each specific run to its cells.
SCP_data <- Reduce(function(x,y) merge(x,y, add.cell.ids = c(x@project.name,y@project.name)), rds_objects)

# check that the size is correct
for (rds_object in rds_objects){
  print(dim(rds_object))
}

print(dim(SCP_data))
```

Removal of big variables used for data merging:

```{r remove_objects, inlude=FALSE}

# remove run objects
rm(rds_object, rds_objects, run_dirs)
```

Saving raw merged data:

```{r save_raw_data, include=FALSE}

saveRDS(SCP_data, file = "~/Documents/SC_Prostate/Data/SC_Prostate_raw.rds")
```

# 3. Visualization of QC metrics.

```{r load_raw_data, include = FALSE}

# SCP_data <- readRDS(file = "~/Documents/SC_Prostate/Data/SC_Prostate_raw.rds")
```

Checking the QC metrics of the combined dataset.

```{r QC_visualization, fig.height=6, fig.width=10}

# Visualize QC metrics as a violin plot
VlnPlot(SCP_data, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

# Zoom in on nFeature_RNA violin plot.
VlnPlot(SCP_data, features = "nFeature_RNA", ncol = 1) + ylim(0, 2500) + NoLegend()

# Zoom in on nCount_RNA violin plot.
VlnPlot(SCP_data, features = "nCount_RNA", ncol = 1) + ylim(0, 25000) + NoLegend()

# Visualize relationships in metadata to detect outliers with FeatureScatter function
plot1 <- FeatureScatter(SCP_data, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatterGradient(SCP_data, feature1 = "nCount_RNA", feature2 = "nFeature_RNA", gradient = "percent.mt", upper.limit = 100, scale.colors = "viridis")
plot1 + plot2
```

We can appreciate the group of cells with low counts and features that we have observed previously during the QC of the individual batches. We kept them in order to analyse its origin. As we will see later, this group of cells cluster together.

Visualizing metrics by sample:

```{r sample_QC, fig.height=6, fig.width=10}

# Visualize QC metrics as a violin plot
VlnPlot(SCP_data, features = "nFeature_RNA", group.by = "Sample_Name") + NoLegend()
VlnPlot(SCP_data, features = "nCount_RNA", group.by = "Sample_Name") + NoLegend()
VlnPlot(SCP_data, features = "percent.mt", group.by = "Sample_Name") + NoLegend()
```

Apparently, run2 and run4 have higher minimum counts than run1 and run3. This could led to batch effect that will be corrected in later steps. Despite this, samples and runs appear to have a pretty similar distribution of QC metrics.

## 3.1. Removal of low cell count samples.

To avoid low-informative samples, we establish a threshold of at least 100 cells for a sample to be considered for the analysis.

```{r calculate_cells_sample}

cells_per_sample <- SCP_data@meta.data %>% group_by(Sample_Name) %>% summarise(Cells = n())
cells_per_sample
```

```{r plot_cells_sample, fig.height=6, fig.width=10}

ggplot(cells_per_sample, aes(x=Sample_Name, y=Cells, fill = Sample_Name)) +
  geom_bar(stat = "identity", color = "black") +
  geom_hline(aes(yintercept=100), linetype="dashed", color="black") +
  theme_classic() + 
  theme(legend.position="none",
        axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)
        )
  
```

```{r filter_samples}

filteredSamples <- cells_per_sample %>% filter(Cells >= 100) %>% select(Sample_Name) %>% unlist(.)
SCP_data <- subset(SCP_data, Sample_Name %in% filteredSamples)
SCP_data
```

# 4. Normalize and remove batch effect.

Batch effect is considered the presence of certain unique characteristics within a specific dataset that make it differentiable from the rest when merging all the data. As we are merging data from different sequencing runs, variations in the samples processing and further cell manipulation and sequencing could induce those batch effect.

Seurat offers methods to integrate data from different batches (different sequencing runs, different samples, etc.). In v5, a new method called IntegrateLayers they easily perform the whole pipeline in a few steps. The first of them is to split the data according to the source of batch effect (in our case by run).

There are several methods for integration, but according to some sources (https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1850-9), the best ones are Harmony, Seurat's CCA and scVII (requires setting up a conda environment and installing reticulate, see documentation). We choose Harmony for being well-integrated with Seruat, performing pretty well with low computational effort, and better respecting the data original structure when compared against Seurat's CCA.

```{r split_data}

# The split method splits the given slot according to the indicated information.
SCP_data[["RNA"]] <- split(SCP_data[["RNA"]], f = SCP_data$orig.ident)
```

# 4.1. Normalize and apply data integration.

It is necessary to normalize the data before integration. According to maintainers (https://github.com/satijalab/seurat/issues/4811, https://github.com/satijalab/seurat/issues/7585) it doesn't mind to normalize before or after splitting.

We now run normalization, PCA and the integration.

```{r norm_batch_effect, warning = FALSE}

SCP_data <- SCP_data %>% 
  NormalizeData(., scale.factor = 1e6, assay = "RNA", verbose = FALSE) %>% # scale.factor = 1e6 means we are using CPMs.
  FindVariableFeatures(., nfeatures = 3000, assay = "RNA", verbose = FALSE) %>%
  ScaleData(., assay = "RNA", vars.to.regress = "percent.mt", verbose = FALSE) %>% # regress out % mito.
  RunPCA(., assay = "RNA", verbose = FALSE) %>%
  IntegrateLayers(., method = HarmonyIntegration, verbose = FALSE) # Integration.
```

If we visualize the PCA plot, we appreciate a bigger overlap between runs.

```{r fig.height=6, fig.width=10}

Idents(SCP_data) <- "orig.ident"
DimPlot(SCP_data, reduction = "pca")
```

## 4.2. Set dimensionality.

While the elbow appears around the 15th PC, we extend the included PCs up to the 20th following the general single-cell analysis guidelines.

```{r elbwol_plot, fig.height=6, fig.width=10}

ElbowPlot(SCP_data, ndims = 50) # choose 20
```

The heatmap and the feature plot give us information about the significance and composition of each PC.

```{r PCA_heatmap, fig.height=15, fig.width=10}

DimHeatmap(SCP_data, dims = 1:20, balanced = TRUE)
```

```{r PCA_top_features, fig.height=30, fig.width=15}

VizDimLoadings(SCP_data, dims = 1:20, reduction = "pca")
```

# 5. Clustering.

After a meticulous clustering resolution fine-tuning a value of 1 was chosen as an appropriate resolution for the clustering.

It is important to notice that a new reduction is created called "harmony". We need to indicate this as the reduction to use in the clustering.

```{r clustering, warning = FALSE}

resolutions = c(0.2, 0.4, 1)

SCP_data <- SCP_data %>% 
  FindNeighbors(., dims = 1:20, reduction = "harmony", verbose = FALSE) %>%
  FindClusters(., resolution =  resolutions, verbose = FALSE) %>% # Default resolution = 0.8
  RunUMAP(., dims = 1:20, reduction = "harmony", verbose = FALSE)
```

## 5.1. Cluster QC.

Cluster QC shows homogeneous characteristics across the clusters with the exception of cluster 22 (abnormal low number of features), which are the group of low feature cells that we have observed in the previous analysis. It is very interesting how they are clustered together, and if that specific group of cells give inconsistent results, because they have been isolated in their own cluster we can move forward just ignoring or removing it.

```{r cluster_QC, fig.height=6, fig.width=10}

# Visualize QC metrics as a violin plot
VlnPlot(SCP_data, features = "nFeature_RNA") + NoLegend()
VlnPlot(SCP_data, features = "nCount_RNA") + NoLegend()
VlnPlot(SCP_data, features = "percent.mt") + NoLegend()

# Zoom in on nFeature_RNA violin plot.
VlnPlot(SCP_data, features = "nFeature_RNA", ncol = 1) + ylim(0, 2500) + NoLegend()

# Visualize relationships in metadata to detect outliers with FeatureScatter function
plot1 <- FeatureScatter(SCP_data, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(SCP_data, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 + NoLegend() + plot2 + NoLegend()
```

## 5.2. Check batch effect with UMAP.

We can use the UMAP plot to visualize the distribution of each sequencing run across the clusters. This allows to detect the remains of batch effect artifacts that causes the clustering of cells coming from the same run.

We can see that the cells from the different batches are equally distributed across the clusters, meaning that the data integration performed with Harmony has been successful.

```{r batch_effect_plot, fig.height=10, fig.width=20}

p1 <- DimPlot(SCP_data, reduction = "umap", label = TRUE, repel = TRUE, label.size = 4) + theme(axis.title = element_text(size = 15), legend.text = element_text(size = 10), axis.text = element_text(size = 10)) + guides(colour = guide_legend(override.aes = list(size = 3))) + NoLegend()

p2 <- DimPlot(SCP_data, reduction = "umap", label = FALSE, group.by = "orig.ident") + theme(axis.title = element_text(size = 15), legend.text = element_text(size = 10), axis.text = element_text(size = 10)) + guides(colour = guide_legend(override.aes = list(size = 3)))

p1 + p2
```

# 6. Annotation.

## 6.1. By-cell annotation.

Cell-by-cell annotation identifies 30 different cell types based on the HumanPrimaryCellAtlas, which distribution pretty much coincides with the cluster layout.

```{r annotation}

# Using celldex and SingleR packages.
# Download reference data from celldex.
reference <- celldex::HumanPrimaryCellAtlasData()

# Join layers before conversion. SingleR uses the RNA assay.
merged.SCexp <- JoinLayers(SCP_data, assay = "RNA") 

# Convert Seurat object into a SingleCellExperiment Object for SingleR input.
merged.SCexp <- as.SingleCellExperiment(merged.SCexp)

SingleR.annotation <- SingleR(test = merged.SCexp, ref = reference, assay.type.test = "logcounts", labels = reference$label.main, num.threads = parallel::detectCores())

SCP_data[["cell.labels"]] <- SingleR.annotation$labels
```

```{r annotation_vis_cell, fig.height=6, fig.width=10}

DimPlot(SCP_data, reduction = "umap", group.by = "cell.labels", label = FALSE) + theme(axis.title = element_text(size = 15), legend.text = element_text(size = 10), axis.text = element_text(size = 10)) + guides(colour = guide_legend(override.aes = list(size = 3)))
```

## 6.2. By-cluster annotation.

We create our own composite cluster names by adding the cluster annotation to the cluster number provided by Seurat. Thus, what we see is an estimation of the cell type that better fits the genetic landscape of each cluster.

```{r cluster_annotation}

# Obtaining a vector containing the cluster of each cell in order.

# Get the factor contained in the SeuratObject with all this information.
clusters <- SCP_data@meta.data[["seurat_clusters"]]

# The cluster information for each cell is contain as a factor which levels coincide with the total number of clusters found by FindClusters(). An approach to transform this factor into a character vector is the following:
# Obtain the list of clusters with levels(clusters). This outputs a character vector containing the levles of the factor. After that, we use the factor itself as an index to access the levels vector. When using a factor as an index, R does not use the labels itself (which in this case are string, so if used as indexes would cause an error), but the internal numeric index the factor contains. That way, for each cluster label associated with a cell in the factor, we access its numeric index and map it to the levels vectors (which coincides), thus obtaining each cell label as an unique character value. Each cell label is then storage as a character (the as.character is added as a control method since SingleR only admits strings as labels) in a vector. The vector contains the cluster label for each cell as a character value in the same order as each cell appears in the dataset, so the by-cluster annotation doesn't assign the cells to an incorrect cluster.
clusters <- as.character(levels(clusters)[clusters])

# reference <- HumanPrimaryCellAtlasData()

# # Join layers before conversion. SingleR uses the RNA assay.
# merged.SCexp <- JoinLayers(SCP_data, assay = "RNA") 
# merged.SCexp <- as.SingleCellExperiment(SCP_data)

# We input the cluster vector using the clusters parameter.
SingleR.annotation <- SingleR(test = merged.SCexp, ref = reference, assay.type.test = "logcounts", labels = reference$label.main, clusters = clusters, num.threads = parallel::detectCores())

SCP_data[["cluster.labels"]] <- SingleR.annotation$labels
```

```{r set_ids}

# We composite the cluster name. That way when 2 clusters' names are the same Seurat doesn't merge the labels.

# Get clusters levels accessing the SeuratObject variable as a df and then accessing the df as a column.
cluster_number <- levels(SCP_data[["seurat_clusters"]][1, ])

# Get annotation labels.
cluster_annotation <- SingleR.annotation$labels

# Since cluster levels and labels are in the same order, we composite the new names using paste0 (sort of equivalent to fstrings in python).
new.clusters.ids <- paste0(cluster_number, "-", cluster_annotation)

# Add names to each value of the clusters id vector so Seurat can take it as a valid input for RenameIdents.
names(new.clusters.ids) <- levels(SCP_data)
SCP_data <- RenameIdents(SCP_data, new.clusters.ids)
SCP_data[["cell.cluster.labels"]] <- Idents(SCP_data)
```

```{r cluster_annotation_vis, fig.height=6, fig.width=10}

DimPlot(SCP_data, reduction = "umap", label = TRUE, repel = TRUE, label.size = 4) + theme(axis.title = element_text(size = 15), legend.text = element_text(size = 10), axis.text = element_text(size = 10)) + guides(colour = guide_legend(override.aes = list(size = 3)))
```


```{r save_processed_data, include=FALSE}

saveRDS(SCP_data, file = "~/Documents/SC_Prostate/Data/SC_Prostate_processed.rds")
```

# 7. Find cluster markers.

The next step is to make a bona fide assignment of the cellular identity of each cluster. For that, we will perform a differential expression analysis of each cluster against the rest of the data set to identify cell markers.

According to the benchmark performed by Sonenson, C. & Robison, D. (2018), the top method for DE with scRNA-seq data is MAST using CPMs (Counts Per Million) and taking into account the Cell Detection Rate (CDR). CDR is described as the proportion of genes expressed in a single cell.

We consider a valid marker those DE genes that are expressed positively in at least 50% of the cells and has a minimum fold change of 1.5.

We also try to correct the batch effect in the analysis using the latent.vars option of FindAllMarkers.

```{r include = FALSE}

# # Load data
# SCP_data <- readRDS("~/Documents/SC_Prostate/Data/SC_Prostate_processed.rds")
```

```{r, results = 'hide'}

# Join the per-run splitted layers.
SCP_data <- JoinLayers(SCP_data, assay = "RNA")

# Calculate Cell Detection Rate (CDR) with our own function (code available in a separate file).
SCP_data <- CalculateCDR(SCP_data)

cluster_markers <- FindAllMarkers(object = SCP_data,
                                  test.use = "MAST", # Use MAST as DE method.
                                  latent.vars = c("orig.ident", "CDR"), # Correct by CDR and batch effect.
                                  assay = "RNA",
                                  slot = "data", # Use log-transformed CPMs.
                                  only.pos = TRUE, # Cell markers can only be positive markers.
                                  min.pct = 0.5, # We consider that a marker gene should be expressed in at least 50% of the cells.
                                  logfc.threshold = log2(1.5), # Filter by FC threshold.
                                  verbose = FALSE
)
```

```{r include = FALSE}

# Save to csv.
write.csv(cluster_markers, "~/Documents/SC_Prostate/Output/cluster_markers_res=1.csv", row.names = TRUE)
```

# 8. Perform GSEA analysis of the clusters.

In order to make the task of identifying the identity of each cluster easier, we will perform a GSEA enrichment analysis for all the genes of each of the clusters that are expressed at least on 50% of the cells. We will use the C8 dataset of GSEA's MSigDB, containing curated cell identity markers obtained from single-cell experiments.

We use the FindAllMarkers function to easily obtain the fold change for the desired genes. The use of the Wilcoxon test has no impact, it is just for optimizing computation time since Seurat uses the presto package for this analysis. In this case, we won't filter out negatively DE genes or apply a FC threshold since using all the available information from the differential expression analysis increases the power of the GSEA.

Nevertheless, the GSEA output will only contain cell identities based on positive markers (entries with a positive Normalized Enrichment Score or NES) and we will keep only results with a p-adj below 1e-6.

```{r warning = FALSE}

# Get DE genes.
cluster_markers <- FindAllMarkers(object = SCP_data,
                                  test.use = "wilcox", # Use Wilcoxon as DE method as it is well-optimized with presto.
                                  assay = "RNA",
                                  slot = "data", # Use log-transformed CPMs.
                                  # We need all genes for the enrichment analysis:
                                  only.pos = FALSE,
                                  min.pct = 0.5, # We keep the pct threshold to avoid underrepresented genes.
                                  logfc.threshold = 0, # No FC threshold.
                                  verbose = FALSE
)

# Load the pathways database from GSEA.
reference_markers <- fgsea::gmtPathways(gmt.file = "~/Documents/SC_Prostate/Data/cell_markers_GSEA.gmt")

# Perform GSEA with our own function (code available in a separate file)
result_list <- scGSEAmarkers(cluster_markers, reference_markers, padj.threshold = 1e-6, only.pos = TRUE, workers = parallel::detectCores())
```

```{r include = FALSE}

# Export each cluster information as a single data.frame:
path <- "~/Documents/SC_Prostate/Output/GSEA_markers_res=1/"

for (cluster in names(result_list)) {
  df <- result_list[[cluster]]
  df <- apply(df,2,as.character) # Transform each column to character to avoid conflicts with lists.
  filename <- paste0(path, cluster, ".csv")
  write.csv(df, file = filename, row.names = FALSE)
  cat("Exported dataframe", cluster, "to", filename, "\n")
}
```

# 9. Finding cluster enriched in T samples.

We want to see which cluster are enriched in tumoral cells. First, we need to calculate the proportion of cells from the different groups (T, N, NAT and AT) on each cluster.

```{r T_enrichment_calc, fig.height=40, fig.width=10}

# Sumarize() performs an operation over a row of a dataframe, for example mean() or count n().

# Slots are dataframes, so it can be used by dplyr.
prop_cell_by_sample <- SCP_data@meta.data %>% group_by(Sample_Name, cell.cluster.labels) %>% 
  summarise(n = n()) %>% # For each cluster, group the cells from the same sample together and count them.
  ungroup() %>% group_by(Sample_Name) %>%
  mutate(freq = n / sum(n)) %>% # Group them now by sample, add up the total number of cells from that sample (regardless of the cluster they belong to). Then, divide each n value (number of cells of a sample in a certain cluster), obtaining which fraction of the total cells of that given type is present on each cluster.
  left_join(SCP_data@meta.data %>% select(Sample_Name, Sample_Group) %>% unique()) # Add metadata info available for the data, select only the Sample_Name and Sample_Group fields and delete duplicates.
```

We focus on comparing the confirmed tumoral (T) and healthy (N) samples.

```{r}

# Filter only the desired groups "N" and "T"
prop_cell_by_sample_filtered <- prop_cell_by_sample %>%
  filter(Sample_Group %in% c("N", "T"))
```

We apply the Wilcoxon test to find out significant differences between sample groups.

```{r boxplot, fig.height=20, fig.width=20}

ggboxplot(prop_cell_by_sample_filtered, x = "Sample_Group", y = "freq",
          color = "Sample_Group",  palette = "jco",
          add = "jitter") +
facet_wrap(~cell.cluster.labels, scales = "free", nrow = 6) +
theme(legend.position = "none") +
xlab("Sample Group") + ylab("Frequency") +
stat_compare_means(aes(label = ..p.signif..), label.x = 1.5)
```


Session Info.

```{r session_info}

sessionInfo()
```
